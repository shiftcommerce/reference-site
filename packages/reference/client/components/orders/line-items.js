import { PureComponent } from 'react'
import Link from 'next/link'
import t from 'typy'

// Lib
import { penceToPounds } from '../../lib/pence-to-pounds'

// Objects
import Image from '../../objects/image'

class LineItems extends PureComponent {
  renderTotal (lineItem) {
    const total = t(lineItem, 'pricing.line_inc_tax').safeObject

    return (
      <div className='c-order-history__line-items_grid_item4'>
        <div className='c-order-history__line-items_total-wrapper'>
          <a className='c-order-history__line-items_total'>&pound;{ penceToPounds(total) }</a>
        </div>
      </div>
    )
  }

  renderImage (lineItem) {
    if (lineItem.imageUrl) {
      return (
        <Link href={`/slug?slug=${lineItem.slug}`} as={`${lineItem.canonical_path}`} key={lineItem.slug}>
          <Image className='c-order-history__line-items_image' src={lineItem.imageUrl} alt={lineItem.title} key={lineItem.slug} aria-label={lineItem.title} />
        </Link>
      )
    } else {
      return (
        <Image className='c-order-history__line-items_image' src='/static/placeholder.png' alt={lineItem.title} key={lineItem.slug} aria-label={lineItem.title} />
      )
    }
  }

  renderTitle (lineItem) {
    return (
      <div className='c-order-history__line-items_grid_item2'>
        <div className='c-order-history__line-items_details'>
          <h5 className='c-order-history__line-items_details_title u-bold'>
            { lineItem.sku }
          </h5>
          <div className='c-order-history__line-items_details_sku'>
            <span>
              { lineItem.sku }
            </span>
          </div>
        </div>
      </div>
    )
  }

  renderParams (lineItem) {
    const deliveryMethod = t(lineItem, 'shipping_method.label').safeObject
    const customerName = t(lineItem, 'shipping_address.name').safeObject

    return (
      <>
        <div className='c-order-history__line-items_grid_item3'>
          <div className='c-order-history__line-items_params u-bold'>
            <span>Quantity: </span>
          </div>
          <div className='c-order-history__line-items_params u-bold'>
            <span>Delivery: </span>
          </div>
          <div className='c-order-history__line-items_params u-bold'>
            <span>Address: </span>
          </div>
        </div>
        <div className='c-order-history__line-items_grid_item3'>
          <span className='c-order-history__line-items_space'>{ lineItem.quantity }</span>
          <span className='c-order-history__line-items_space'>{ deliveryMethod }</span>
          <span className='c-order-history__line-items_space'>{ customerName }</span>
        </div>
      </>
    )
  }

  renderLineItems () {
    const { items } = this.props

    return (
      items.map((lineItem, index) =>
        <div className='c-order-history__line-items_grid' key={lineItem.sku}>
          <div className='c-order-history__line-items_grid_item1'>
            { this.renderImage(lineItem) }
          </div>
          { this.renderTitle(lineItem) }
          { this.renderParams(lineItem) }
          { this.renderTotal(lineItem) }
        </div>
      )
    )
  }

  render () {
    return (
      <div className='c-order-history__line-items'>
        <h3 className='c-order-history__line-items_title'>Products</h3>
        { this.renderLineItems() }
      </div>
    )
  }
}

export default LineItems
