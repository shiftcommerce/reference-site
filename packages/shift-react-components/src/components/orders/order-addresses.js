// Libraries
import React, { PureComponent } from 'react'

export class OrderAddresses extends PureComponent {
  render () {
    return (
      <>
        { this.props.addresses.map((address) => {
          return (
            <div className='c-order-history__address' key={address.id}>
              <p className='u-bold'>{ address.name }</p>
              <p>{ address.company }</p>
              { address.lines.map((line => { return <p>{line}</p> })) }
              <p>{ address.city }</p>
              <p>{ address.state }</p>
              <p>{ address.postcode }</p>
              <p>{ address.country }</p>
            </div>
          )
        })}
      </>
    )
  }
}
