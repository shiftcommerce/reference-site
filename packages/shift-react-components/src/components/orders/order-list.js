// Libraries
import React, { PureComponent, Fragment } from 'react'
import format from 'date-fns/format'
import t from 'typy'

// Lib
import { fulfillmentStatus } from '../../lib/fulfillment-status'
import { penceToPounds } from '../../lib/pence-to-pounds'

export class OrderList extends PureComponent {
  constructor (props) {
    super(props)

    this.headers = props.headers || ['Order No.', 'Order Date', 'Cost', 'Shipping Status', '']
  }

  /**
  * Render the order details
  * @param  {Object} order
  * @param  {Object} orderDate
  * @param  {Object} total
  * @return {string} - HTML markup for the component
  */
  renderOrderDetails (order) {
    const { updateCurrentOrder } = this.props

    const orderDate = format(new Date(order.placed_at), 'D MMM YYYY')
    const orderTotal = penceToPounds(t(order, 'pricing.total_inc_tax').safeObject)

    return [
      {
        label: this.headers[0],
        value: order.reference
      },
      {
        label: this.headers[1],
        value: orderDate
      },
      {
        label: this.headers[2],
        value: `${order.pricing.currency}${orderTotal}`
      },
      {
        label: this.headers[3],
        value: fulfillmentStatus(order.fulfillment_status)
      },
      {
        label: this.headers[4],
        value: <a href={`/account/orders/${order.reference}`} className='o-button o-button--primary' onClick={(event) => updateCurrentOrder(event, order.reference)}>View Details</a>
      }
    ].map((column, columnIndex) => {
      return (
        <div
          key={columnIndex}
          className='c-order-history-table__col'
        >
          { column.label && <span className='c-order-history-table__label'>{ column.label }</span> }
          { column.value }
        </div>
      )
    })
  }

  /**
  * Render the header for the order list
  * @param  {Object} order
  * @param  {Object} orderDate
  * @param  {Object} total
  * @return {string} - HTML markup for the component
  */
  renderOrderHeader (order, orderDate, total) {
    return (
      <Fragment>
        <h4>{ orderDate } - { order.reference } - Â£{ total }</h4>
        <label htmlFor={order.reference} className='c-order-history__header-label' />
        <input type='checkbox' id={order.reference} />
        <span className='c-order-history__arrow' />
      </Fragment>
    )
  }

  renderTableHeader () {
    return this.headers.map((column, columnIndex) => {
      return (
        <div
          key={columnIndex}
          className='c-order-history-table__col c-order-history-table__col--header'
        >
          { column }
        </div>
      )
    })
  }

  render () {
    const { orders } = this.props

    return (
      <div className='c-order-history-table'>
        <div className='c-order-history-table__row c-order-history-table__row--header'>
          { this.renderTableHeader() }
        </div>
        { orders.data.map((order) => {
          return (
            <div
              key={order.id}
              className='c-order-history-table__row c-order-history-table__row--body'
            >
              { this.renderOrderDetails(order) }
            </div>
          )
        }) }
      </div>
    )
  }
}
