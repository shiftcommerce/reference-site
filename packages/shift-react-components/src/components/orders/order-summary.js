// Libraries
import React, { Component } from 'react'
import Sticky from 'react-stickyfill'

// Lib
import { decimalPrice } from '../../lib/decimal-price'
import Config from '../../lib/config'

// Objects
import link from '../../objects/link'
import { Image } from '../../objects/image'

const Link = Config.get().Link || link

export class OrderSummary extends Component {
  /**
  * Render line item image for the order summary
  * @param  {Object} lineItem
  * @return {string} - HTML markup for the component
  */
  renderLineItemImage (lineItem) {
    if (lineItem.imageUrl) {
      return (
        <Link href={`/slug?slug=${lineItem.slug}`} as={`${lineItem.canonical_path}`} >
          <Image src={lineItem.imageUrl} alt={lineItem.title} aria-label={lineItem.title} />
        </Link>
      )
    } else {
      return (
        <Image src='/static/placeholder.png' alt={lineItem.title} aria-label={lineItem.title} />
      )
    }
  }

  /**
  * Render the line item title for the order summary
  * @param  {Object} lineItem
  * @return {string} - HTML markup for the component
  */
  renderLineItemTitle (lineItem) {
    return (
      <>
        <p className='c-order__summary-line-item-details-title u-bold'>
          { lineItem.title }
        </p>
        <p className='c-order__summary-line-item-details-sku'>
          { lineItem.sku }
        </p>
      </>
    )
  }

  /**
  * Render the line-item meta-data for the order summary
  * @param  {Object} _lineItem
  * @return {string} - HTML markup for the component
  */
  renderLineItemParams (_lineItem) {
    // @TODO: to update this with actual metadata
    const metaData = { colour: 'Grey', size: 'L', quantity: 1 }
    return (
      Object.keys(metaData).map((key, index) =>
        <div className='c-order__summary-line-item-details-param' key={key}>
          <div className='c-order__summary-line-item-details-param-key'>
            <p>{ key }:</p>
          </div>
          <div className='c-order__summary-line-item-details-param-value'>
            <p>{ metaData[key] }</p>
          </div>
        </div>
      )
    )
  }

  /**
  * Render the line items for the order summary
  * @param  {Object} order
  * @return {string} - HTML markup for the component
  */
  renderLineItems (order) {
    return (
      order.line_items.map((lineItem) =>
        <div className='c-order__summary-line-item' key={lineItem.sku}>
          <div className='c-order__summary-line-item-image'>
            { this.renderLineItemImage(lineItem) }
          </div>

          <div className='c-order__summary-line-item-details'>
            { this.renderLineItemTitle(lineItem) }
            { this.renderLineItemParams(lineItem) }
          </div>
        </div>
      )
    )
  }

  /**
  * Render the total prices for the order summary
  * @param  {Object} order
  * @return {string} - HTML markup for the component
  */
  renderOrderTotals (order) {
    return (
      <Sticky>
        <div aria-label='Order total summary' className='u-sticky'>
          <div>
            <dl aria-label='Order subtotal'>
              <dt>Total products:</dt>
              <dd>&pound;{ decimalPrice(order.total_inc_tax - order.shipping_total) }</dd>
            </dl>
            <dl aria-label='Order shipping costs'>
              <dt>Shipping costs:</dt>
              <dd>&pound;{ order.shipping_total }</dd>
            </dl>
            <dl aria-label='Order total' className='u-bold'>
              <dt>TOTAL:</dt>
              <dd>&pound;{ decimalPrice(order.total_inc_tax) }</dd>
            </dl>
            <dl>
              <dt>* Including VAT</dt>
            </dl>
          </div>
        </div>
      </Sticky>
    )
  }

  render () {
    const { order } = this.props

    return (
      <div className='c-order__summary'>
        <div className='c-order__summary-header'>
          <div className='c-order__summary-header-title'>
            <h2>Your Order</h2>
          </div>

          <div className='c-order__summary-header-total'>
            <h2>&pound;{ decimalPrice(order.total_inc_tax) }</h2>
          </div>
        </div>

        { this.renderLineItems(order) }

        <div className='c-order__summary-totals'>
          { this.renderOrderTotals(order) }
        </div>
      </div>
    )
  }
}
