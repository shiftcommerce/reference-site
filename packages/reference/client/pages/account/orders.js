// Libraries
import { connect } from 'react-redux'

// Pages etc
import { algoliaReduxWrapper } from '@shiftcommerce/shift-next/src/lib/algolia-redux-wrapper'
import { AccountOrdersPage } from '@shiftcommerce/shift-next/src/pages/account/orders'
import { MyAccountLayout } from '@shiftcommerce/shift-next/src/layouts/my-account'

// Config
import menu from '../../config/myaccount-menu.js'

const layout = {
  component: MyAccountLayout,
  props: { menu }
}

function mapStateToProps ({ orders }) { return { orders, layout } }

export default algoliaReduxWrapper(connect(mapStateToProps)(AccountOrdersPage), AccountOrdersPage)
