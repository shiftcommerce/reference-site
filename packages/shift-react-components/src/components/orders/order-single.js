// Libraries
import React, { PureComponent, Fragment } from 'react'
import format from 'date-fns/format'

// Lib
import { penceToPounds } from '../../lib/pence-to-pounds'

// Components
import ShippingAddresses from './shipping-addresses'
import { LineItems } from '../cart/line-items'

import Link from '../../objects/link'

export class OrderSingle extends PureComponent {
  /**
   * Calculate the subtotal based on each line item price
   * @param  {Object} order
   * @return {Number}
   */
  calculateSubTotal (order) {
    return order.line_items.reduce(
      (accumulator, currentValue) => accumulator + currentValue.pricing.line_total_inc_tax, 0
    )
  }

  /**
   * Render the total shipping costs in the order list
   * @param  {Object} order
   * @return {string} - HTML markup for the component
   */
  calculateShippingTotal (order) {
    return order.shipping_methods.reduce(
      (accumulator, currentValue) => accumulator + currentValue.price, 0
    )
  }

  /**
   * Get the tax amount based on the totals
   * @param  {Object} order
   * @return {Number}
   */
  calculateTax (order) {
    return order.pricing.total_inc_tax - order.pricing.total_ex_tax
  }

  /**
   * Reformat the line items to they'll work with the <LineItems /> component
   * @param  {Array} lineItems
   * @return {Array}
   */
  formatLineItems (lineItems) {
    return lineItems.map((lineItem, lineItemIndex) => {
      return {
        id: lineItemIndex,
        item: {
          sku: lineItem.sku,
          title: 'item title',
          picture_url: lineItem.image_urls[0],
          product: {
            title: 'product title',
            slug: 'slug'
          }
        },
        unit_quantity: lineItem.quantity,
        total: (lineItem.pricing.line_total_inc_tax / 100),
        sub_total: (lineItem.pricing.line_total_inc_tax / 100)
      }
    })
  }

  /**
   * Render each part of the order items
   * @param  {String}                 options.header
   * @param  {String|Number|Element0} options.body
   * @return {String} - HTML markup for the component
   */
  renderItem ({ header, body }) {
    return (
      <Fragment>
        { header && <h5 className='c-order-single__header'>{ header }</h5> }
        { body && <div className='c-order-single__body'>{ body }</div> }
      </Fragment>
    )
  }

  render () {
    const { order } = this.props
    const lineItems = this.formatLineItems(order.line_items)

    const items = [{
      header: 'Order No.',
      body: order.reference
    }, {
      header: 'Order Date',
      body: format(new Date(order.placed_at), 'MMM D, YYYY')
    }, {
      header: 'Cost',
      body: `&pound;${penceToPounds(order.pricing.total_inc_tax)}`
    }, {
      header: 'Shipping Method',
      body: order.shipping_methods[0].label
    }, {
      header: 'Billing Address',
      body: <ShippingAddresses addresses={order.shipping_addresses} />
    }, {
      header: 'Shipping Address',
      body: <ShippingAddresses addresses={order.shipping_addresses} />
    }, {
      header: 'Currency',
      body: order.pricing.currency
    }]

    return (
      <Fragment>
        <div className='c-order-history__nav'>
          <h2>Order { order.reference }</h2>
        </div>
        <div className='c-order-single'>
          { items.map((item, itemIndex) => {
            return (
              <div key={itemIndex} className='c-order-single__item'>
                { this.renderItem(item) }
              </div>
            )
          }) }
          <div className='c-order-single__item c-order-single__item--full'>
            <h5 className='c-order-single__header'>Order Summary</h5>
            <div className='c-order-single__body'>
              <div className='o-grid-container'>
                <div className='o-col-1-7'>
                  <LineItems
                    lineItems={lineItems}
                    lineItemsCount={lineItems.length}
                  />
                </div>
                <div className='o-col-8-13'>
                  <div className='c-cart-summary'>
                    <dl aria-label='Subtotal'>
                      <dt> Subtotal: </dt>
                      <dd> &pound;{ penceToPounds(this.calculateSubTotal(order)) } </dd>
                    </dl>
                    <dl aria-label='Taxes'>
                      <dt> UK VAT 20% (Included in Price): </dt>
                      <dd> &pound;{ penceToPounds(this.calculateTax(order)) } </dd>
                    </dl>
                    <dl aria-label='Shipping'>
                      <dt> Shipping Total: </dt>
                      <dd> &pound;{ penceToPounds(this.calculateShippingTotal(order)) } </dd>
                    </dl>
                    <dl aria-label='Total' className='c-cart-summary__total'>
                      <dt> Total: </dt>
                      <dd> &pound;{ penceToPounds(order.pricing.total_inc_tax) } </dd>
                    </dl>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div className='c-order-single__item c-order-single__item--full'>
            <div className='c-order-single__body'>
              <Link href='/account/orders' className='o-button o-button--primary'>View all orders</Link>
            </div>
          </div>

        </div>
      </Fragment>
    )
  }
}
