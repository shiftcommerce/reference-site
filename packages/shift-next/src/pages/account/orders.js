// Libraries
import React, { Component } from 'react'

// Components
import { AccountOrders } from '@shiftcommerce/shift-react-components/src/components/account/orders'

// Actions
import { getCustomerOrders } from '../../actions/account-actions'

export class AccountOrdersPage extends Component {
  constructor (props) {
    super(props)

    this.state = {
      currentOrderRef: null
    }

    this.fetchOrders = this.fetchOrders.bind(this)
    this.updateCurrentOrder = this.updateCurrentOrder.bind(this)
  }

  static async getInitialProps (appContext) {
    // extracts page number from query or set default
    return {
      pageNumber: (appContext.query.page || 1)
    }
  }

  updateCurrentOrder (event, currentOrderRef = null) {
    event.preventDefault()
    this.setState({ currentOrderRef })
  }

  /*
   * Retrieves customer orders
   * @return {Object} - orders response
   */
  fetchOrders () {
    this.props.dispatch(getCustomerOrders(this.props.pageNumber))
  }

  render () {
    const { layout, orders, pageNumber } = this.props
    const { currentOrderRef } = this.state
    const Layout = layout.component

    return (
      <Layout {...layout.props}>
        <AccountOrders
          currentOrderRef={currentOrderRef}
          fetchOrders={this.fetchOrders}
          orders={orders}
          pageNumber={pageNumber}
          pagePath={'/account/orders'}
          updateCurrentOrder={this.updateCurrentOrder}
        />
      </Layout>
    )
  }
}
