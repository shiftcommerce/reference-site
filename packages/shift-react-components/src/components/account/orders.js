// Libraries
import React, { Component, Fragment } from 'react'

// Lib
import Config from '../../lib/config'
import { Loading } from '../../objects/loading'
import { OrderList } from '../../components/orders/order-list'
import { OrderSingle } from '../../components/orders/order-single'

export class AccountOrders extends Component {
  constructor (props) {
    super(props)

    this.Link = Config.get().Link
  }

  componentDidMount () {
    this.props.fetchOrders()
  }

  /**
   * Render the list of orders, or the individual order
   * @param  {Object} orders
   * @return {string} - HTML markup for the component
   */
  renderOrders (orders) {
    const {
      currentOrderRef,
      fetchOrders,
      pageNumber,
      pagePath,
      updateCurrentOrder
    } = this.props

    if (orders.data.length === 0) {
      return (<p>No previous orders found.</p>)
    }

    if (currentOrderRef) {
      const currentOrder = orders.data.find((order) => {
        return order.reference === currentOrderRef
      })

      return (
        <OrderSingle
          order={currentOrder}
          updateCurrentOrder={updateCurrentOrder}
        />
      )
    }

    return (
      <Fragment>
        <div className='c-order-history__nav'>
          <h2>Order History</h2>
        </div>
        <OrderList
          orders={orders}
          pageNumber={pageNumber}
          pagePath={pagePath}
          updateCurrentOrder={updateCurrentOrder}
        />
      </Fragment>
    )
  }

  render () {
    const { orders, orders: { loading } } = this.props

    return (
      <div className='c-order-history'>
        { loading ? <Loading /> : this.renderOrders(orders) }
      </div>
    )
  }
}
