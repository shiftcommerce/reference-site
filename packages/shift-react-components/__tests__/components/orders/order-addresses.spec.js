// Libraries
import React from 'react'

// Component
import { OrderAddresses } from '../../../src/components/orders/order-addresses'

// Fixtures
import orders from '../../fixtures/orders'

test('renders correctly', () => {
  // Arrange
  const shippingAddresses = orders.data[0].shipping_addresses

  // Act
  const wrapper = mount(
    <OrderAddresses addresses={shippingAddresses} />
  )

  // Assert
  expect(wrapper).toMatchSnapshot()
  expect(wrapper).toIncludeText(shippingAddresses[0].name)
  expect(wrapper).toIncludeText(shippingAddresses[0].company)
})
