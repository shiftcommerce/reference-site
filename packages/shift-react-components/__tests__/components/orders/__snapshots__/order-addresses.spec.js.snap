// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<OrderAddresses
  addresses={
    Array [
      Object {
        "city": "Wakefield",
        "company": "COMPANY",
        "country": "GB",
        "id": "f52d57e1-5960-4bd6-8cd4-3f6b856b365d_1",
        "lines": Array [
          "1 Scott Drive",
          "Crofton",
        ],
        "name": "Callum Barratt",
        "postcode": "WF4 1TD",
        "state": "West Yorkshire",
      },
    ]
  }
>
  <div
    className="c-order-history__address"
    key="f52d57e1-5960-4bd6-8cd4-3f6b856b365d_1"
  >
    <p
      className="u-bold"
    >
      Callum Barratt
    </p>
    <p>
      COMPANY
    </p>
    <p
      key="0"
    >
      1 Scott Drive
    </p>
    <p
      key="1"
    >
      Crofton
    </p>
    <p>
      Wakefield
    </p>
    <p>
      West Yorkshire
    </p>
    <p>
      WF4 1TD
    </p>
    <p>
      GB
    </p>
  </div>
</OrderAddresses>
`;
